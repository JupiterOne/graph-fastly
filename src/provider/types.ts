export type FastlyAccount = {
  id: string;
  created_at: string;
  updated_at: string;
  name: string;
  owner_id: string;
  deleted_at: null;
  pricing_plan: string;
  billing_contact_id: string | null;
  postal_address: string | null;
  billing_ref: string | null;
  phone_number: string;
  pricing_plan_id: string | string;
  can_upload_vcl: boolean;
  billing_network_type: string;
  can_configure_wordpress: boolean;
  can_reset_passwords: boolean;
  force_2fa: boolean;
  force_sso: boolean;
  has_account_panel: boolean;
  has_api_key: boolean;
  has_improved_events: boolean;
  has_improved_ssl_config: boolean;
  has_openstack_logging: boolean;
  has_pci: boolean;
  has_pci_passwords: boolean;
  ip_whitelist: string;
  logentries_provisioned_by: string | null;
  max_tokens_per_user: number;
  rate_limit: number;
  readonly: boolean;
  requires_support_email: boolean;
  san_domains: number;
};

// https://docs.fastly.com/en/guides/configuring-user-roles-and-permissions#user-roles-and-what-they-can-do
export enum FastlyUserRole {
  USER = 'user',
  BILLING = 'billing',
  ENGINEER = 'engineer',
  SUPERUSER = 'superuser',
}

export type FastlyUser = {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
  customer_id: string;
  require_new_password: boolean;
  role: FastlyUserRole;
  login: string;
  deleted_at: string | null;
  locked: boolean;
  two_factor_auth_enabled: boolean;
  two_factor_setup_required: boolean;
  limit_services: boolean;
  email_hash: string;
};

export type FastlyToken = {
  id: string;
  user_id: string;
  customer_id: string;
  name: string;
  last_used_at: string;
  created_at: string;
  expires_at: string | null;
  ip: string;
  user_agent: string;
  sudo_expires_at: string | null;
  scopes: string[];
  scope: string;
  services: [];
  service_id: string | null;
};

export type FastlyService = {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
  customer_id: string;
  type: string;
  comment: string;
  version: number;
  versions: object[];
};

export type FastlyServiceBackend = {
  name: string;
  version: number;
  weight: number;
  hostname: string | null;
  port: number;
  address: string;
  ipv4: string;
  ipv6: string | null;
  comment: string;
  auto_loadbalance: boolean;
  between_bytes_timeout: number;
  client_cert: string | null;
  connect_timeout: number;
  error_threshold: number;
  first_byte_timeout: number;
  healthcheck: string | null;
  request_condition: string;
  locked: boolean;
  min_tls_version: string | null;
  max_tls_version: string | null;
  max_conn: number;
  override_host: string | null;
  shield: string;
  ssl_ciphers: string | null;
  ssl_cert_hostname: string | null;
  ssl_hostname: string | null;
  ssl_client_cert: string | null;
  ssl_ca_cert: string | null;
  ssl_check_cert: boolean;
  ssl_client_key: string | null;
  ssl_sni_hostname: string | null;
  use_ssl: boolean;
  service_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
};

export type FastlyServiceDomain = {
  name: string;
  locked: boolean;
  comment: string;
  version: number;
  service_id: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
};
