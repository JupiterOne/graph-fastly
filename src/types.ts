import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Customer ID obtained from https://manage.fastly.com/account/company
   */
  customerId: string;

  /**
   * The API Token created from https://manage.fastly.com/account/personal/tokens
   */
  apiToken: string;
}
