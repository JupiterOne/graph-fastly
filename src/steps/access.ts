import {
  createDirectRelationship,
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  getTime,
  Entity,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../provider/client';
import { IntegrationConfig } from '../types';

export async function fetchUsers({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const accountEntity = (await jobState.getData(
    `fastly-account:${instance.config.customerId}`,
  )) as Entity;

  const apiClient = createAPIClient(instance.config);

  await apiClient.iterateUsers(async (data) => {
    const userEntity = createIntegrationEntity({
      entityData: {
        source: data,
        assign: {
          _type: 'fastly_user',
          _class: 'User',
          _key: `fastly-user:${data.id}`,
          id: data.id,
          displayName: data.name || data.id,
          name: data.name,
          username: data.login,
          login: data.login,
          role: data.role,
          admin: data.role === 'superuser',
          mfaEnabled: data.two_factor_auth_enabled,
          mfaSetupRequired: data.two_factor_setup_required,
          locked: data.locked,
          active: !(data.deleted_at || data.locked),
          createdOn: getTime(data.created_at),
          updatedOn: getTime(data.updated_at),
          deletedOn: getTime(data.deleted_at),
        },
      },
    });

    await Promise.all([
      jobState.addEntity(userEntity),
      jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: accountEntity,
          to: userEntity,
        }),
      ),
    ]);
  });
}

export async function fetchTokens({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  await apiClient.iterateTokens(async (data) => {
    const tokenEntity = createIntegrationEntity({
      entityData: {
        source: data,
        assign: {
          _type: 'fastly_api_token',
          _class: 'AccessKey',
          _key: `fastly-api-token:${data.id}`,
          id: data.id,
          displayName: data.name || data.id,
          name: data.name,
          ip: data.ip,
          userAgent: data.user_agent,
          scope: data.scope,
          scopes: data.scopes,
          services: data.services,
          userId: data.user_id,
          createdOn: getTime(data.created_at),
          expiresOn: getTime(data.expires_at),
          sudoExpiresOn: getTime(data.sudo_expires_at),
          lastUsedOn: getTime(data.last_used_at),
          lastUsedAt: data.ip,
          lastUsedBy: data.user_agent,
        },
      },
    });

    const fromEntity = data.user_id
      ? {
          _key: `fastly-user:${data.user_id}`,
          _type: 'fastly_user',
          _class: 'User',
        }
      : {
          _key: `fastly-account:${instance.config.customerId}`,
          _type: 'fastly-account',
          _class: 'Account',
        };

    await Promise.all([
      jobState.addEntity(tokenEntity),
      jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: fromEntity,
          to: tokenEntity,
        }),
      ),
    ]);
  });
}

export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-users',
    name: 'Fetch Users',
    entities: [
      {
        resourceName: 'User',
        _type: 'fastly_user',
        _class: 'User',
      },
    ],
    relationships: [
      {
        _type: 'fastly_account_has_user',
        _class: RelationshipClass.HAS,
        sourceType: 'fastly_account',
        targetType: 'fastly_user',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchUsers,
  },
  {
    id: 'fetch-tokens',
    name: 'Fetch API Tokens',
    entities: [
      {
        resourceName: 'API Token',
        _type: 'fastly_api_token',
        _class: 'AccessKey',
      },
    ],
    relationships: [
      {
        _type: 'fastly_account_has_api_token',
        _class: RelationshipClass.HAS,
        sourceType: 'fastly_account',
        targetType: 'fastly_api_token',
      },
      {
        _type: 'fastly_user_has_api_token',
        _class: RelationshipClass.HAS,
        sourceType: 'fastly_user',
        targetType: 'fastly_api_token',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchTokens,
  },
];
