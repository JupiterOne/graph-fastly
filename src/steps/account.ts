import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  getTime,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../types';
import { createAPIClient } from '../provider/client';

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const data = await apiClient.getAccountDetails();

  const accountEntity = createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _type: 'fastly_account',
        _class: 'Account',
        _key: `fastly-account:${data.id}`,
        id: data.id,
        displayName: data.name || data.id,
        name: data.name,
        address: data.postal_address,
        phone: data.phone_number,
        mfaEnabled: data.force_2fa,
        mfaEnforced: data.force_2fa,
        ssoEnforced: data.force_sso,
        readOnly: data.readonly,
        ownerId: data.owner_id,
        ipWhitelist: data.ip_whitelist,
        rateLimit: data.rate_limit,
        createdOn: getTime(data.created_at),
        updatedOn: getTime(data.updated_at),
        deletedOn: getTime(data.deleted_at),
      },
    },
  });

  await Promise.all([
    jobState.addEntity(accountEntity),
    jobState.setData(accountEntity._key, accountEntity),
  ]);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Account',
        _type: 'fastly_account',
        _class: 'Account',
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
