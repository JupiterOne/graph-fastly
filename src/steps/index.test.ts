import {
  createMockStepExecutionContext,
  setupRecording,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../types';
import { fetchTokens, fetchUsers } from './access';
import { fetchAccountDetails } from './account';
import { fetchServices } from './services';

const DEFAULT_CUSTOMER_ID = 'abc';
const DEFAULT_API_TOKEN = 'xyz';

const integrationConfig: IntegrationConfig = {
  customerId: process.env.CUSTOMER_ID || DEFAULT_CUSTOMER_ID,
  apiToken: process.env.API_TOKEN || DEFAULT_API_TOKEN,
};

jest.setTimeout(10000 * 2);

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  await fetchAccountDetails(context);
  await fetchUsers(context);
  await fetchTokens(context);
  await fetchServices(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Account'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'fastly_account' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name'],
    },
  });

  expect(
    context.jobState.collectedEntities.filter((e) => e._class.includes('User')),
  ).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'fastly_user' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name'],
    },
  });

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('AccessKey'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['AccessKey'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'fastly_api_token' },
        userId: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['userId'],
    },
  });

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Service'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Service'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'fastly_service' },
        name: { type: 'string' },
        version: { type: 'number' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'version'],
    },
  });
});
