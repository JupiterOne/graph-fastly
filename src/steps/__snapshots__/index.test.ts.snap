// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "billing_contact_id": null,
            "billing_network_type": "public",
            "billing_ref": null,
            "can_configure_wordpress": false,
            "can_reset_passwords": true,
            "can_upload_vcl": true,
            "created_at": "2020-08-27T13:32:45Z",
            "deleted_at": null,
            "force_2fa": false,
            "force_sso": false,
            "has_account_panel": true,
            "has_api_key": false,
            "has_improved_events": false,
            "has_improved_ssl_config": false,
            "has_openstack_logging": true,
            "has_pci": false,
            "has_pci_passwords": false,
            "id": "6Bfjl1cR8HV2hPuQykFxJa",
            "ip_whitelist": "0.0.0.0/0",
            "logentries_provisioned_by": null,
            "max_tokens_per_user": 100,
            "name": "JupiterOne",
            "owner_id": "1hJNx2XBX94T64w931jVBp",
            "phone_number": "9195228915",
            "postal_address": null,
            "pricing_plan": "developer",
            "pricing_plan_id": "6HW4welznAHVnMCJhY9QdX",
            "rate_limit": 1000,
            "readonly": false,
            "requires_support_email": false,
            "san_domains": 5,
            "updated_at": "2020-08-27T13:32:45Z",
          },
        },
      ],
      "_type": "fastly_account",
      "address": null,
      "createdOn": 1598535165000,
      "deletedOn": undefined,
      "displayName": "JupiterOne",
      "id": "6Bfjl1cR8HV2hPuQykFxJa",
      "ipWhitelist": "0.0.0.0/0",
      "mfaEnabled": false,
      "mfaEnforced": false,
      "name": "JupiterOne",
      "ownerId": "1hJNx2XBX94T64w931jVBp",
      "phone": "9195228915",
      "rateLimit": 1000,
      "readOnly": false,
      "ssoEnforced": false,
      "updatedOn": 1598535165000,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "fastly-user:1hJNx2XBX94T64w931jVBp",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2020-08-27T13:32:45Z",
            "customer_id": "6Bfjl1cR8HV2hPuQykFxJa",
            "deleted_at": null,
            "email_hash": "c60680fa19ac66cbb41f6898b99bd367",
            "id": "1hJNx2XBX94T64w931jVBp",
            "limit_services": false,
            "locked": false,
            "login": "erkang.zheng@jupiterone.com",
            "name": "Erkang Zheng",
            "require_new_password": false,
            "role": "superuser",
            "two_factor_auth_enabled": false,
            "two_factor_setup_required": false,
            "updated_at": "2020-08-27T13:34:47Z",
          },
        },
      ],
      "_type": "fastly_user",
      "active": true,
      "admin": true,
      "createdOn": 1598535165000,
      "deletedOn": undefined,
      "displayName": "Erkang Zheng",
      "id": "1hJNx2XBX94T64w931jVBp",
      "locked": false,
      "login": "erkang.zheng@jupiterone.com",
      "mfaEnabled": false,
      "mfaSetupRequired": false,
      "name": "Erkang Zheng",
      "role": "superuser",
      "updatedOn": 1598535287000,
      "username": "erkang.zheng@jupiterone.com",
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "fastly-api-token:4sPcKZXcDLkMy3aoVwWeqT",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2020-08-27T13:35:02Z",
            "customer_id": "6Bfjl1cR8HV2hPuQykFxJa",
            "expires_at": null,
            "id": "4sPcKZXcDLkMy3aoVwWeqT",
            "ip": "99.29.119.164",
            "last_used_at": "2020-08-28T21:32:03Z",
            "name": "manage.fastly.com browser session",
            "scope": "global",
            "scopes": Array [
              "global",
            ],
            "service_id": null,
            "services": Array [],
            "sudo_expires_at": null,
            "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
            "user_id": "1hJNx2XBX94T64w931jVBp",
          },
        },
      ],
      "_type": "fastly_api_token",
      "createdOn": 1598535302000,
      "displayName": "manage.fastly.com browser session",
      "expiresOn": undefined,
      "id": "4sPcKZXcDLkMy3aoVwWeqT",
      "ip": "99.29.119.164",
      "lastUsedAt": "99.29.119.164",
      "lastUsedBy": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
      "lastUsedOn": 1598650323000,
      "name": "manage.fastly.com browser session",
      "scope": "global",
      "scopes": Array [
        "global",
      ],
      "services": Array [],
      "sudoExpiresOn": undefined,
      "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
      "userId": "1hJNx2XBX94T64w931jVBp",
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "fastly-api-token:6cw4NpEw9wWDeDkhDFzAEq",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2020-08-29T14:35:06Z",
            "customer_id": "6Bfjl1cR8HV2hPuQykFxJa",
            "expires_at": null,
            "id": "6cw4NpEw9wWDeDkhDFzAEq",
            "ip": "99.29.119.164",
            "last_used_at": "2020-08-29T19:22:25Z",
            "name": "manage.fastly.com browser session",
            "scope": "global",
            "scopes": Array [
              "global",
            ],
            "service_id": null,
            "services": Array [],
            "sudo_expires_at": "2020-08-29T18:27:19Z",
            "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
            "user_id": "1hJNx2XBX94T64w931jVBp",
          },
        },
      ],
      "_type": "fastly_api_token",
      "createdOn": 1598711706000,
      "displayName": "manage.fastly.com browser session",
      "expiresOn": undefined,
      "id": "6cw4NpEw9wWDeDkhDFzAEq",
      "ip": "99.29.119.164",
      "lastUsedAt": "99.29.119.164",
      "lastUsedBy": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
      "lastUsedOn": 1598728945000,
      "name": "manage.fastly.com browser session",
      "scope": "global",
      "scopes": Array [
        "global",
      ],
      "services": Array [],
      "sudoExpiresOn": 1598725639000,
      "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
      "userId": "1hJNx2XBX94T64w931jVBp",
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "fastly-api-token:7B77tg0OSKAnEHtEzwcaJp",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2020-08-29T14:43:33Z",
            "customer_id": "6Bfjl1cR8HV2hPuQykFxJa",
            "expires_at": null,
            "id": "7B77tg0OSKAnEHtEzwcaJp",
            "ip": "99.29.119.164",
            "last_used_at": "2020-08-29T20:55:20Z",
            "name": "J1",
            "scope": "global:read",
            "scopes": Array [
              "global:read",
            ],
            "service_id": null,
            "services": Array [],
            "sudo_expires_at": null,
            "user_agent": "axios/0.19.2",
            "user_id": "1hJNx2XBX94T64w931jVBp",
          },
        },
      ],
      "_type": "fastly_api_token",
      "createdOn": 1598712213000,
      "displayName": "J1",
      "expiresOn": undefined,
      "id": "7B77tg0OSKAnEHtEzwcaJp",
      "ip": "99.29.119.164",
      "lastUsedAt": "99.29.119.164",
      "lastUsedBy": "axios/0.19.2",
      "lastUsedOn": 1598734520000,
      "name": "J1",
      "scope": "global:read",
      "scopes": Array [
        "global:read",
      ],
      "services": Array [],
      "sudoExpiresOn": undefined,
      "userAgent": "axios/0.19.2",
      "userId": "1hJNx2XBX94T64w931jVBp",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "comment": "",
            "created_at": "2020-08-29T14:35:22Z",
            "customer_id": "6Bfjl1cR8HV2hPuQykFxJa",
            "id": "3EgDEph9cJ5gjzE2WHgDCw",
            "name": "FastTest",
            "type": "vcl",
            "updated_at": "2020-08-29T14:41:22Z",
            "version": 2,
            "versions": Array [
              Object {
                "active": false,
                "comment": "",
                "created_at": "2020-08-29T14:35:22Z",
                "deleted_at": null,
                "deployed": false,
                "locked": false,
                "number": 1,
                "service_id": "3EgDEph9cJ5gjzE2WHgDCw",
                "staging": false,
                "testing": false,
                "updated_at": "2020-08-29T17:16:36Z",
              },
              Object {
                "active": false,
                "comment": "",
                "created_at": "2020-08-29T15:03:03Z",
                "deleted_at": null,
                "deployed": false,
                "locked": true,
                "number": 2,
                "service_id": "3EgDEph9cJ5gjzE2WHgDCw",
                "staging": false,
                "testing": false,
                "updated_at": "2020-08-29T17:15:14Z",
              },
            ],
          },
        },
      ],
      "_type": "fastly_service",
      "category": Array [
        "infrastructure",
      ],
      "createdOn": 1598711722000,
      "description": "",
      "displayName": "FastTest",
      "id": "3EgDEph9cJ5gjzE2WHgDCw",
      "name": "FastTest",
      "updatedOn": 1598712082000,
      "version": 2,
    },
    Object {
      "_class": Array [
        "ApplicationEndpoint",
      ],
      "_key": "fastly-service-backend:null:13.249.40.78:443",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "address": "13.249.40.78",
            "auto_loadbalance": false,
            "between_bytes_timeout": 10000,
            "client_cert": null,
            "comment": "",
            "connect_timeout": 1000,
            "created_at": "2020-08-29T14:40:12Z",
            "deleted_at": null,
            "error_threshold": 0,
            "first_byte_timeout": 15000,
            "healthcheck": null,
            "hostname": null,
            "ipv4": "13.249.40.78",
            "ipv6": null,
            "locked": true,
            "max_conn": 200,
            "max_tls_version": null,
            "min_tls_version": null,
            "name": "Host 1",
            "override_host": null,
            "port": 443,
            "request_condition": "",
            "service_id": "3EgDEph9cJ5gjzE2WHgDCw",
            "shield": "",
            "ssl_ca_cert": null,
            "ssl_cert_hostname": null,
            "ssl_check_cert": false,
            "ssl_ciphers": null,
            "ssl_client_cert": null,
            "ssl_client_key": null,
            "ssl_hostname": null,
            "ssl_sni_hostname": null,
            "updated_at": "2020-08-29T15:03:03Z",
            "use_ssl": true,
            "version": 2,
            "weight": 100,
          },
        },
      ],
      "_type": "fastly_service_backend",
      "address": "13.249.40.78",
      "autoLoadbalance": false,
      "betweenBytesTimeout": 10000,
      "comment": "",
      "connectTimeout": 1000,
      "createdAt": "2020-08-29T14:40:12Z",
      "createdOn": 1598712012000,
      "deletedOn": undefined,
      "description": "",
      "displayName": "Host 1",
      "errorThreshold": 0,
      "firstByteTimeout": 15000,
      "ipAddress": "13.249.40.78",
      "ipv4": "13.249.40.78",
      "locked": true,
      "maxConn": 200,
      "name": "Host 1",
      "port": 443,
      "requestCondition": "",
      "serviceId": "3EgDEph9cJ5gjzE2WHgDCw",
      "shield": "",
      "sslCheckCert": false,
      "updatedAt": "2020-08-29T15:03:03Z",
      "updatedOn": 1598713383000,
      "useSsl": true,
      "version": 2,
      "weight": 100,
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "fastly-service:5pqQ8jn0OPmRASXrwpmiID",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "comment": "",
            "created_at": "2020-08-27T13:35:17Z",
            "customer_id": "6Bfjl1cR8HV2hPuQykFxJa",
            "id": "5pqQ8jn0OPmRASXrwpmiID",
            "name": "Unnamed Service",
            "type": "vcl",
            "updated_at": "2020-08-27T13:35:17Z",
            "version": 1,
            "versions": Array [
              Object {
                "active": false,
                "comment": "",
                "created_at": "2020-08-27T13:35:17Z",
                "deleted_at": null,
                "deployed": false,
                "locked": false,
                "number": 1,
                "service_id": "5pqQ8jn0OPmRASXrwpmiID",
                "staging": false,
                "testing": false,
                "updated_at": "2020-08-27T13:35:17Z",
              },
            ],
          },
        },
      ],
      "_type": "fastly_service",
      "category": Array [
        "infrastructure",
      ],
      "createdOn": 1598535317000,
      "description": "",
      "displayName": "Unnamed Service",
      "id": "5pqQ8jn0OPmRASXrwpmiID",
      "name": "Unnamed Service",
      "updatedOn": 1598535317000,
      "version": 1,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa",
      "_key": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa|has|fastly-user:1hJNx2XBX94T64w931jVBp",
      "_toEntityKey": "fastly-user:1hJNx2XBX94T64w931jVBp",
      "_type": "fastly_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-user:1hJNx2XBX94T64w931jVBp",
      "_key": "fastly-user:1hJNx2XBX94T64w931jVBp|has|fastly-api-token:4sPcKZXcDLkMy3aoVwWeqT",
      "_toEntityKey": "fastly-api-token:4sPcKZXcDLkMy3aoVwWeqT",
      "_type": "fastly_user_has_api_token",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-user:1hJNx2XBX94T64w931jVBp",
      "_key": "fastly-user:1hJNx2XBX94T64w931jVBp|has|fastly-api-token:6cw4NpEw9wWDeDkhDFzAEq",
      "_toEntityKey": "fastly-api-token:6cw4NpEw9wWDeDkhDFzAEq",
      "_type": "fastly_user_has_api_token",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-user:1hJNx2XBX94T64w931jVBp",
      "_key": "fastly-user:1hJNx2XBX94T64w931jVBp|has|fastly-api-token:7B77tg0OSKAnEHtEzwcaJp",
      "_toEntityKey": "fastly-api-token:7B77tg0OSKAnEHtEzwcaJp",
      "_type": "fastly_user_has_api_token",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa",
      "_key": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa|has|fastly-service:3EgDEph9cJ5gjzE2WHgDCw",
      "_toEntityKey": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw",
      "_type": "fastly_account_has_service",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw",
      "_key": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw|has|fastly-service-backend:null:13.249.40.78:443",
      "_toEntityKey": "fastly-service-backend:null:13.249.40.78:443",
      "_type": "fastly_service_has_backend",
      "displayName": "HAS",
    },
    Object {
      "_class": "CONNECTS",
      "_key": "fastly-service-backend:null:13.249.40.78:443|connects|FORWARD:_class=Host,Gateway:ipAddress=13.249.40.78:_class=Host,Gateway:hostname=null",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "skipTargetCreation": true,
        "sourceEntityKey": "fastly-service-backend:null:13.249.40.78:443",
        "targetEntity": Object {
          "_class": Array [
            "Host",
            "Gateway",
          ],
          "hostname": null,
          "ipAddress": "13.249.40.78",
        },
        "targetFilterKeys": Array [
          Array [
            "_class",
            "ipAddress",
          ],
          Array [
            "_class",
            "hostname",
          ],
        ],
      },
      "_type": "fastly_service_backend_connects_host",
      "displayName": "CONNECTS",
    },
    Object {
      "_class": "CONNECTS",
      "_key": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw|connects|FORWARD:_class=DomainRecord:name=jupiterone.com",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw",
        "targetEntity": Object {
          "_class": "DomainRecord",
          "name": "jupiterone.com",
        },
        "targetFilterKeys": Array [
          Array [
            "_class",
            "name",
          ],
        ],
      },
      "_type": "fastly_service_connects_domain_record",
      "createdOn": 1598711735000,
      "deletedOn": undefined,
      "description": "",
      "displayName": "CONNECTS",
      "locked": true,
      "updatedOn": 1598713383000,
    },
    Object {
      "_class": "CONNECTS",
      "_key": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw|connects|FORWARD:_class=DomainRecord:name=securitymanifesto.net",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "fastly-service:3EgDEph9cJ5gjzE2WHgDCw",
        "targetEntity": Object {
          "_class": "DomainRecord",
          "name": "securitymanifesto.net",
        },
        "targetFilterKeys": Array [
          Array [
            "_class",
            "name",
          ],
        ],
      },
      "_type": "fastly_service_connects_domain_record",
      "createdOn": 1598711800000,
      "deletedOn": undefined,
      "description": "",
      "displayName": "CONNECTS",
      "locked": true,
      "updatedOn": 1598713383000,
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa",
      "_key": "fastly-account:6Bfjl1cR8HV2hPuQykFxJa|has|fastly-service:5pqQ8jn0OPmRASXrwpmiID",
      "_toEntityKey": "fastly-service:5pqQ8jn0OPmRASXrwpmiID",
      "_type": "fastly_account_has_service",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "fastly_account",
    "fastly_user",
    "fastly_account_has_user",
    "fastly_api_token",
    "fastly_user_has_api_token",
    "fastly_service",
    "fastly_account_has_service",
    "fastly_service_backend",
    "fastly_service_has_backend",
    "fastly_service_backend_connects_host",
    "fastly_service_connects_domain_record",
  ],
  "numCollectedEntities": 8,
  "numCollectedRelationships": 10,
}
`;
